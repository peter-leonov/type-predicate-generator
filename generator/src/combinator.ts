import { PseudoBigInt } from "typescript";
import { assert, unimplemented } from "./helpers";
import {
  AliasType,
  ArrayType,
  LiteralType,
  ObjectType,
  PrimitiveType,
  TypeModel,
  UnionType,
} from "./model";

/**
 * Main invariant: do not reuse the values produced by the value funtions
 * (value, union, array, object), they each get a unique ID that is
 * used to remember if the value has produced a broken value.
 *
 * If this becomes a problem, the next in line solution is to again
 * use value paths as IDs and store per path state.
 */

/**
 * A unique value used in tests to mark a wrong value that does not match
 * anything esle but itself and cannot be expressed in the subset of the
 * type system supported by the generator.
 *
 * This explicit value covers cases when in a union one member's invalid value
 * is a valid for the other. For example in:
 * ```ts
 * type X = string | number
 * ```
 * the `string` value might choose a `1` as it's invalid value,
 * but `1` is a valid value for the `number` thus the union as a whole
 * never yields an invalid value.
 *
 * If this special value becomes a blocker it should be possible to
 * inspect the union values in the generator and infer a safe invalid value
 * per union.
 */
const invalidValue: any = Symbol("invalidValue");

export type Value =
  | number
  | string
  | null
  | undefined
  | boolean
  | { [key: string]: Value }
  | PseudoBigInt
  | Value[];

export type ValueGenerator<T> = (
  doInvalid: boolean
) => Generator<[boolean, T]>;

export function value<T>(value: T): ValueGenerator<T> {
  return function* (doInvalid: boolean) {
    if (doInvalid) {
      yield [false, invalidValue];
    }
    yield [true, value];
  };
}

export function union<T>(
  members: ValueGenerator<T>[]
): ValueGenerator<T> {
  return function* (doInvalid: boolean) {
    for (const m of members) {
      yield* m(doInvalid);
    }
  };
}

export function array<T>(
  value: ValueGenerator<T>
): ValueGenerator<T[]> {
  return function* (doInvalid: boolean) {
    if (doInvalid) {
      yield [false, invalidValue];
    }
    yield [true, []];
    for (const [isValid, v] of value(doInvalid)) {
      yield [isValid, [v]];
    }
  };
}

export function object<T>(
  obj: Record<string, ValueGenerator<T>>
): ValueGenerator<T> {
  return function* (doInvalid: boolean) {
    if (doInvalid) {
      yield [false, invalidValue];
    }
    yield* rollObject(doInvalid, obj) as Generator<any>;
  };
}

function* rollObject<T>(
  doInvalid: boolean,
  obj: Record<string, ValueGenerator<T>>
): Generator<[boolean, Record<string, T>]> {
  const keys = Object.keys(obj);
  if (keys.length == 0) {
    yield [true, {}];
    return;
  }

  const [key, ...restKeys] = keys;
  assert(key, "at least one property must be present at this point");

  const restObj = pick(obj, restKeys);

  const value = obj[key];
  assert(value);

  if (doInvalid) {
    for (const [_, rest] of rollObject(false, restObj)) {
      yield [false, rest];
    }
  }

  for (const [isValidValue, v] of value(doInvalid)) {
    for (const [isValidRest, rest] of rollObject(
      doInvalid && isValidValue,
      restObj
    )) {
      yield [
        isValidValue && isValidRest,
        {
          [key]: v,
          ...rest,
        },
      ];
    }
  }
}

function pick<T>(
  obj: Record<string, T>,
  arr: string[]
): Record<string, T> {
  return arr.reduce<Record<string, T>>((acc, attr) => {
    if (attr in obj) {
      acc[attr] = obj[attr] as T;
    }
    return acc;
  }, {});
}

export function combineValid(fn: ValueGenerator<Value>) {
  return [...fn(false).map(([isValid, v]) => (assert(isValid), v))];
}

export function combineInvalid(fn: ValueGenerator<Value>) {
  return [
    ...fn(true)
      .filter(([isValid, _]) => !isValid)
      .map(([_, v]) => v),
  ];
}

export function modelToCombinator(
  model: TypeModel
): ValueGenerator<Value> {
  if (model instanceof LiteralType) {
    return value(model.value);
  } else if (model instanceof PrimitiveType) {
    switch (model.primitive) {
      case "string":
        return value("string");
      case "number":
        return union([
          // adding 0 to cover for checks like `!x` where `0` would meant `false`
          value(0),
          // any arbitrary number
          value(42),
        ]);
      case "boolean":
        return union([value(true), value(false)]);
    }
    model.primitive satisfies never;
    unimplemented(
      `primitive type ${model.primitive} is not implemented`
    );
  } else if (model instanceof ObjectType) {
    return object(
      mapObjectValues(model.attributes, modelToCombinator)
    );
  } else if (model instanceof UnionType) {
    return union(model.types.map(modelToCombinator));
  } else if (model instanceof ArrayType) {
    return array(modelToCombinator(model));
  } else if (model instanceof AliasType) {
    unimplemented(
      "AliasType test generation is to be implemented soon"
    );
  }

  model satisfies never;
  unimplemented(
    `model of class ${className(model)} is not implemented`
  );
}

function className(value: unknown): string {
  if (typeof value === "object" && value !== null) {
    return value.constructor.name;
  }
  return "undefined";
}

function mapObjectValues<T, U>(
  src: Record<string, T>,
  f: (v: T) => U
): Record<string, U> {
  const dst: Record<string, U> = {};
  for (var k in src) {
    dst[k] = f(src[k] as any);
  }
  return dst;
}
