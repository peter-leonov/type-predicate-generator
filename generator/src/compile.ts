import ts from "typescript";
import { TypeGuardGenerator } from "./generator";
import { assert, unimplemented } from "./helpers";
import { typeToModel } from "./model";

/**
 * This is required for both `X[]` and `Array<X>` to work properly.
 * Otherwise `checker.isArrayType()` is not gonna work.
 */
const libFileContent = `/// <reference no-default-lib="true"/>
interface Boolean {}
interface Function {}
interface CallableFunction {}
interface NewableFunction {}
interface IArguments {}
interface Number { toExponential: any; }
interface Object {}
interface RegExp {}
interface String { charAt: any; }
interface Array<T> { length: number; [n: number]: T; }
interface ReadonlyArray<T> {}
declare const console: { log(msg: any): void; };`;

export function newVFSProgram(source: string): ts.Program {
  const fileName = "/source.ts";

  const files = new Map<string, string>();
  files.set(fileName, source);

  const libFileName = "/lib/lib.d.ts";
  files.set(libFileName, libFileContent);

  const compilerHost: ts.CompilerHost = {
    fileExists: (fileName) => files.has(fileName),
    getSourceFile: (fileName, options) => {
      const content = files.get(fileName);
      assert(content != undefined, "source files cannot be missing");
      return ts.createSourceFile(fileName, content, options);
    },
    getDefaultLibFileName: () => "/lib/lib.d.ts",
    writeFile: unimplemented,
    getCurrentDirectory: () => "/",
    getDirectories: () => [],
    getCanonicalFileName: (f) => f.toLowerCase(),
    getNewLine: () => "\n",
    useCaseSensitiveFileNames: () => false,
    readFile: (path) => files.get(path),
  };

  return ts.createProgram(
    [fileName],
    {
      target: ts.ScriptTarget.ESNext,
      module: ts.ModuleKind.ESNext,
      strict: true,
      noEmit: true,
      isolatedModules: true,
      // don't autoload all the types from the node_modules/@types
      types: [],
    },
    compilerHost
  );
}

class ProgramError extends Error {}

export function ensureNoErrors(program: ts.Program) {
  const allDiagnostics = ts.getPreEmitDiagnostics(program);
  if (allDiagnostics.length != 0) {
    allDiagnostics.forEach((diagnostic) => {
      var message = ts.flattenDiagnosticMessageText(
        diagnostic.messageText,
        "\n"
      );
      if (!diagnostic.file) {
        console.error(message);
        return;
      }
      var { line, character } =
        diagnostic.file.getLineAndCharacterOfPosition(
          diagnostic.start!
        );
      console.error(
        `${diagnostic.file.fileName} (${line + 1},${
          character + 1
        }): ${message}`
      );
    });

    throw new ProgramError(
      "ts.getPreEmitDiagnostics() found some issues listed in the console / stderr"
    );
  }
}

export function generateFullFileBodyForAllTypes(
  checker: ts.TypeChecker,
  sourceFile: ts.SourceFile,
  importFrom: string
): ts.Statement[] {
  const generator = new TypeGuardGenerator();

  // Walk the tree to search for types
  ts.forEachChild(sourceFile, (node) => {
    if (ts.isTypeAliasDeclaration(node)) {
      let symbol = checker.getSymbolAtLocation(node.name);
      assert(symbol, "type alias declaration must have a symbol");
      const model = typeToModel(
        checker,
        checker.getDeclaredTypeOfSymbol(symbol),
        symbol
      );

      generator.addRootTypeGuardFor(model);
    }

    if (ts.isInterfaceDeclaration(node)) {
      let symbol = checker.getSymbolAtLocation(node.name);
      assert(symbol, "interface declaration must have a symbol");
      const model = typeToModel(
        checker,
        checker.getDeclaredTypeOfSymbol(symbol),
        symbol
      );

      generator.addRootTypeGuardFor(model);
    }
  });

  return generator.getFullFileBody(importFrom);
}
