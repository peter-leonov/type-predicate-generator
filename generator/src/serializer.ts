import ts, { factory } from "typescript";
import { Token, type Value } from "./combinator";
import { unimplemented, unreachable } from "./helpers";

export function valueToNode(value: Value): ts.Expression {
  // built-ins
  if (typeof value === "number") {
    return factory.createNumericLiteral(value);
  } else if (typeof value === "string") {
    return factory.createStringLiteral(value);
  } else if (typeof value === "boolean") {
    return value ? factory.createTrue() : factory.createFalse();
  } else if (typeof value === "undefined") {
    // historically an identifier while having been
    // a built-in sorta constant for ages
    return factory.createIdentifier("undefined");
  } else if (value === null) {
    // historically an identifier while having been
    // a built-in sorta constant for ages
    return factory.createNull();
  } else if (Array.isArray(value)) {
    return factory.createArrayLiteralExpression(
      value.map(valueToNode)
    );
  }

  if (value instanceof Token) {
    unimplemented();
  }
  // TODO: custom types

  // catch-all object
  if (typeof value === "object") {
    return factory.createObjectLiteralExpression(
      Object.entries(value).map(([k, v]) =>
        factory.createPropertyAssignment(
          factory.createIdentifier(k),
          valueToNode(v)
        )
      )
    );
  }

  value satisfies never;
  unreachable(`value ${value} is not supported`);
}
