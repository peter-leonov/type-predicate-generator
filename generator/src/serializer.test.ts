import { expect, test } from "vitest";
import { factory } from "typescript";
import { valueToNode } from "./serializer";
import { nodesToString } from "./compile";
import { type Value } from "./combinator";

function process(value: Value): string {
  return nodesToString("foo", [
    factory.createExpressionStatement(
      factory.createParenthesizedExpression(valueToNode(value))
    ),
  ]);
}

test("number", () => {
  expect(process(42)).toMatchSnapshot();
});

test("string", () => {
  expect(process("foo")).toMatchSnapshot();
});

test("boolean", () => {
  expect(process(true)).toMatchSnapshot();
  expect(process(false)).toMatchSnapshot();
});

test("array", () => {
  expect(process([])).toMatchSnapshot();
  expect(process([1, 2, 3])).toMatchSnapshot();
});

test("object", () => {
  expect(process({})).toMatchSnapshot();
  expect(process({ a: 1, b: 2, c: 3 })).toMatchSnapshot();
});

test("nested", () => {
  expect(process({ a: [1], b: { c: { d: 4 } } })).toMatchSnapshot();
});
