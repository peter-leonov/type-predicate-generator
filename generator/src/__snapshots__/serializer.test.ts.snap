// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array 1`] = `
"([]);
"
`;

exports[`array 2`] = `
"([1, 2, 3]);
"
`;

exports[`boolean 1`] = `
"(true);
"
`;

exports[`boolean 2`] = `
"(false);
"
`;

exports[`nested 1`] = `
"({ a: [1], b: { c: { d: 4 } } });
"
`;

exports[`number 1`] = `
"(42);
"
`;

exports[`object 1`] = `
"({});
"
`;

exports[`object 2`] = `
"({ a: 1, b: 2, c: 3 });
"
`;

exports[`string 1`] = `
"("foo");
"
`;
